ee/desktop/runner/runner.go

```diff
@@ -212,40 +233,92 @@ func (r *DesktopUsersProcessesRunner) Interrupt(interruptError error) {
 		return fmt.Errorf("failed to decode desktop user control data: %w", err)
 	}
 
-	if desktopStatus.Status == "" {
-		return fmt.Errorf("empty desktop status")
+	// Regardless, we will write the buffer data out to a file that can be grabbed by
+	// any desktop user processes, either when they refresh, or when they are spawned.
+	r.writeStatusFile(&buffer)
+
+	if r.processSpawningEnabled && !desktopStatus.Enabled {
+		// Desktop is currently enabled, control server is asking us to disable
+		level.Debug(r.logger).Log("msg", "runner is disabling desktop users processes")
+
+		// Stop automatically spawning desktop user processes
+		r.processSpawningEnabled = false
+
+		// Kill any existing desktop user processes, and we're done here
+		r.killDesktopProcesses()
+		return nil
 	}
 
-	// TODO: TBD how desktop user processes are notified and ingest control data
-	// Below code leverages the HTTP server but I'm thinking it's preferable
-	// to persist the data to a file, and optionally "ping" the server to let it know
-	// an update has occurred. The desktop can read from the file and react on it's own
-	// One advantage here is that the status info can be passed along even if the desktop
-	// server is unavailable (it would read the file on startup)
+	if !r.processSpawningEnabled && desktopStatus.Enabled {
+		// Desktop is currently disabled, control server is asking us to enable
+		level.Debug(r.logger).Log("msg", "runner is enabling desktop users processes")
 
-	// for uid, proc := range r.uidProcs {
-	// 	client := client.New(r.authToken, proc.socketPath)
-	// 	if err := client.SetStatus(desktopStatus.Status); err != nil {
-	// 		level.Error(r.logger).Log(
-	// 			"msg", "error sending status command to desktop process",
-	// 			"uid", uid,
-	// 			"pid", proc.process.Pid,
-	// 			"path", proc.path,
-	// 			"err", err,
-	// 		)
-	// 	}
-	// }
+		// Next execute loop tick will begin spawning desktop user processes
+		r.processSpawningEnabled = true
+		return nil
+	}
+
+	// If we're here:
+	// A) Desktop is currently disabled, and control server wants it to remain disabled
+	// B) Desktop is currently enabled, and control server wants it to remain enabled
+	//
+	// In case A), there will be no process records and thus no servers to talk to
+	// In case B), each process will have a record, and we can use that to talk to its server
+
+	// Tell any running desktop user processes that they should refresh the latest status data
+	for uid, proc := range r.uidProcs {
+		client := client.New(r.authToken, proc.socketPath)
+		if err := client.Refresh(); err != nil {
+			level.Error(r.logger).Log(
+				"msg", "error sending refresh command to desktop process",
+				"uid", uid,
+				"pid", proc.process.Pid,
+				"path", proc.path,
+				"err", err,
+			)
+		}
+	}
+
+	return nil
+}
+
+// writeStatusFile copies desktop status data to a shared file for user processes to access
+func (r *DesktopUsersProcessesRunner) writeStatusFile(data io.Reader) error {
```

Comment:
Why write to a file instead of just sending a POST with the data to the server?

Reply:
My main motivation was to be able to more easily handle status updates when desktop is disabled, or the user processes are otherwise unavailable.

Particularly, when desktop is disabled and a new update makes it enabled, if the only way to send that data is via the client-server, we have to hold on to that update and wait until the process is spawned and the server becomes available. I could have added code to trigger this quickly and not wait for the ticker, and keep the update in-memory until the server is listening, but it would have made things more complex and introduced new edge cases.

By writing it out to a file, we're able to quickly consume the update, and the existing user process lifecycle management code stays largely the same, other than slurping up that file when spawned and when told to refresh.

I'm still writing tests, but I _think_ this will make it easier to test. Also I'm thinking a non-engineer could easily mock desktop menus by modifying the JSON file on disk and refreshing.

Reply:
Thanks for the explanation! That makes sense to me.